### PR TITLE
New implementation of XonshLexer for pygments

### LINKED ISSUE #985: Update pyghooks for @$(...)
I noticed that the `@` doesn't colorize with our new operator.

### PR DESCRIPTION
**Pull request by blahgeek**
This is successor of #1688 , hopefully this would resolve #985 #1663 and maybe #1038 later

BashLexer is removed because that's unnecessary and increases complexity. A large amount of keywords/builtins/syntax in BashLexer is invalid for xonsh's subprocess mode.

Dynamic highlighting for command and path is implemented.

---

Updates:

screen shot 2016-09-12 at 12 00 02
screen shot 2016-09-12 at 11 56 48
screen shot 2016-09-12 at 11 59 12
screen shot 2016-09-12 at 12 00 18
screen shot 2016-09-12 at 12 00 59

### CODE REVIEW

File: tests/test_ptk_highlight.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
+# -*- coding: utf-8 -*-
+"""Test XonshLexer for pygments"""
+
+import pytest
+import os
```
Comment (scopatz) [inline tests/test_ptk_highlight.py]: PEP8 import order, standard library imports should be first.

File: xonsh/pyghooks.py [hunk 2]
```diff
@@ -15,11 +15,12 @@
 load_module_in_background('pkg_resources', debug='XONSH_DEBUG',
                           replacements={'pygments.plugin': 'pkg_resources'})
 
-from pygments.lexer import inherit, bygroups, using, this
+from pygments.lexer import inherit, bygroups, using, this, include
 from pygments.lexers.shell import BashLexer
```
Comment (scopatz) [inline xonsh/pyghooks.py]: Can probably remove this import now, right?

**Comment by scopatz** [comment]
Thanks for this @blahgeek! Looks like were are seeing some test failures, but otherwise have only minor suggestions

**Comment by blahgeek** [comment]
Sure, I am still working on it (more syntax and more tests, and of course, cleaning up the codes, etc.). Just thought would be good to let people know that I am working on this and can get some suggestions. Thanks for your reviewing!

**Comment by scopatz** [comment]
No problem, I think this is definitely the right direction to go in.

**Comment by santagada** [comment]
This is very cool, I was going to open a ticket to implement command highlight everywhere (not only on first token) but I think this will take care of it. Can we highlight invalid commands just like fish and have a token type for commands so they can be a different color from keywords.

**Comment by blahgeek** [comment]
@santagada For now, for the invalid commands in top-level, e.g. `not-a-command xxx`, the lexer would try to parse it as python statement, so it would displayed as `Text`. For the invalid commands in captured subprocess mode, e.g. `$(not-a-command xxx)`, since we can be sure that this should be a command, it would displayed as `Error`.

And sure, I would think more carefully about picking `Keyword` or `Token` or something else for all these tokens.

Thanks for your comment!

**Comment by blahgeek** [comment]
The failed tests on windows should be gone after/if #1702 is merged

**Comment by gforsyth** [comment]
Hey @blahgeek, I just merged #1703 if you want to rebase/merge master

**Comment by blahgeek** [comment]
@scopatz @santagada I think it's almost ready now, if you want to try it.

**Comment by santagada** [comment]
testing it... seems pretty cool so far.

**Comment by santagada** [comment]
first note:
while typing a complete path to a command eg: /bin/sh or /usr/bin/banner you will see that the hightlight is using binary search (looking for x_ok) and not doing the path hightlight. I think it should be path if dir and only do the binary if it is a file and x_ok. I think there is a way to get both information with one stat call.

So the behaviour I was expecting was file path highlight until the last letter of the file is entered and then it turns to the cmd highlight token.

**Comment by blahgeek** [comment]
@santagada Check it out now!

**Comment by santagada** [comment]
@blahgeek still not working for me eg:

```sh
$ /             # / is white
$ /usr       # /usr is white
$ /usr/bi   # still all white
$ /usr/bin # now bin is blue and the rest is white
$ /usr/bin/banner # now everything is blue
```

maybe I didn't update it properly, but I did remove the amalgman file and all .pyc files before installing it again, and I see 5194406dac3a5f18fc6415e037e3c3e944f060ad as the last commit

**Comment by santagada** [comment]
I think the way it is right now is great and better than what was there before... so I would still merge it and let this small things for a next PR

**Comment by laerus** [comment]
hey @santagada try also removing `$XONSH_DATA_DIR/xonsh_code_cache` there are some related bugs with that after installing.

**Comment by santagada** [comment]
one suggestion for tests:

```
def test_first_path():
     HERE = os.path.abspath(os.path.dirname(__file__))
     test_dir = os.path.join(HERE, 'xonsh-test-highlight-path')
     if not os.path.exists(test_dir):
         os.mkdir(test_dir)
     check_token(HERE, [(Name.Constant, HERE)])
     check_token('{}-xxx'.format(HERE), [
                                                (Text,
                                                 '{}-xxx'.format(HERE))
                                                ])

     check_token('./{}'.format(test_dir), [(Name.Constant, './{}'.format(test_dir))])
     check_token('./{}-xxx'.format(test_dir), [
                                                (Text,
                                                 '{}-xxx'.format(test_dir))
                                                ])
```

if you want I can add a bunch of tests and do a PR for this myself

**Comment by santagada** [comment]
@laerus thanks but it didn't help.

**Comment by melund** [comment]
@santagada .. I think we shouldn't make files and folder in the test directory. Pytest can give you a temporary directory with the `tmpdir` fixture. I think your example would look something like this

``` 
def test_first_path(tmpdir):
with tmpdir.as_cwd():
    HERE = tmpdir.strpath
    testdir = tmpdir.mkdir("xonsh-test-highlight-path").strpath
    ....
```

**Comment by blahgeek** [comment]
@santagada currently, the first word would highlight as 1. command, if it's a file and executable, or 2. path, if it's a directory and `$AUTO_CD` is set to True, or 3. text, if non of above conditions apply.

Are you suggesting that a part of path to a executable file should also be highlighted as command?

**Comment by blahgeek** [comment]
@melund aha thanks! `tmpdir` fixture, that's really helpful!

**Comment by gforsyth** [comment]
This looks really good to me.  Amazing work, @blahgeek and it will feel good to close out all of the Lexer/Pygments issues.  

I think this is ready to go in, pending CI finishing up.  Any objections, @laerus, @melund?

**Comment by gforsyth** [comment]
Thanks a ton, @blahgeek !

**Pull request merged**